docs/js-performance.1
." vim: set syn=nroff
.TH js-performance 1 "February 2018" "js-performance v1.17.3"

.SH NAME
js-performance - A JavaScript profiling tool and collection of profiling modules and benchmarks.

.SH SYNOPSIS
js-performance  [-hjlqv]
                [-i \fI\,iterations\fR] [-m \fI\,magnitude\fR] [-p \fI\,time-precision,memory-precision\fR] [-u \fI\,time-unit,memory-unit\fR]
                [\fI\,profile1\fR \fI\,profile2\fR ...]

.SH DESCRIPTION
\fI\,js-performance\fR is a profiling tool and collection of profiling modules and benchmarks. It provides both a CLI and a JavaScript library for use within custom applications.

.SH COMMANDS
\fI\,js-performance\fR supports two distinct commands.

If run with the \fB\,-l\fR or \fB\,--list\fR option, a list of available performance profiles is printed to the terminal.

If run without this option, either all available performance profiles or the ones specified after the command are executed and results printed to the terminal.

.SH ARGUMENTS
To run only specific profiles provide them as arguments separated by spaces. e.g.:
.PP
.nf
.RS
js-performance guards loops
.RE
.fi
.PP
To retrieve a list of available profiles, see the \fI\,--list\fR option.

.SH OPTIONS
-h, --help
.RS
display usage information.
.RE

-i \fI\,iterations\fR, --iterations=\fI\,iterations\fR
.RS
set the number of iterations per profiled function, i.e. how often each profiled function is executed. DEFAULT: 1000
.RE

-j, --json
.RS
output results in JSON format.
.RE

-l, --list
.RS
print a list of available performance profiles.
.RE

-m \fI\,magnitude\fR, --magnitude=\fI\,magnitude\fR
.RS
set the magnitude of test data provided to each profiled function, i.e. how many items each function has to process. DEFAULT: 1000
.RE

-p \fI\,time-precision,memory-precision\fR, --precision=\fI\,time-precision,memory-precision\fR
.RS
set the floating point precision to use for results, i.e. the number of decimal places to use for time and memory output. Separate time and memory precision by comma. DEFAULT: 4
.RE

-u \fI\,time-unit,memory-unit\fR, --unit=\fI\,time-unit,memory-unit\fR
.RS
set the unit used for results. Separate time and memory unit by comma.

Available units:
.RS
.IP auto .2i
automatically convert between milli- and microseconds. DEFAULT
.IP ms
milliseconds
.IP \[u03BC]s
microseconds. ATTENTION: the first letter is an actual mu.
.IP B
.br
Byte
.IP KB
Kilobyte
.IP MB
Megabyte
.RE
.RE

-q, --quiet
.RS
print results only.
.RE

-v, --verbose
.RS
print verbose output.
.RE

.SH LIBRARY
Refer to <https://github.com/haensl/js-performance/blob/master/docs/lib.md> for information on how to use the js-performance library in your own applications.

.SH BUGS
Report bugs to <https://github.com/haensl/js-performance/issues>

.SH AUTHOR
Written by Hans-Peter Dietz and Fabian Maschewski.



CHANGELOG.md
### 1.18.1
* [#67: Update intro.](https://github.com/haensl/js-profiler/issues/67)

### 1.18.0
* [#65: Rename to js-profiler.](https://github.com/haensl/js-performance/issues/65)

### 1.17.3
* [#62: Remove global `requireModule()`](https://github.com/haensl/js-performance/issues/62)

### 1.17.2
* [#59: Update readme.](https://github.com/haensl/js-performance/issues/59)

### 1.17.1
* [#24: Handling VERBOSITY in profiles is like inline styles.](https://github.com/haensl/js-performance/issues/24)

### 1.17.0
* [#55: Publish node/v8 and package version in results.](https://github.com/haensl/js-performance/issues/55)

### 1.16.0
* [#54: Add npmignore.](https://github.com/haensl/js-performance/issues/54)

### 1.15.0
* [#30: Measure memory consumption.](https://github.com/haensl/js-performance/issues/30)
* Refactor function execution and profiling into profiler module.
* Add units for memory measurement.
* Enable precision option to distinguish between time and memory precision.
* Enable unit option to distinguish between time and memory unit.

### 1.14.0
* [#31: Minimize pollution for accurate measurements.](https://github.com/haensl/js-performance/issues/31)
* Adjust clock handling.
* Move high resolution time conversion to units module.
* Fix pollution in guards profile.

### 1.13.0
* [#49: Add man page.](https://github.com/haensl/js-performance/issues/49)
* Update dependencies to fix security vulnerability in brace-expansion < 1.1.7

### 1.12.0
* [#28: Output minimum, maximum, average runtime for each test.](https://github.com/haensl/js-performance/issues/28)

### 1.11.0
* [#46: Create npm module and binary.](https://github.com/haensl/js-performance/issues/46)
* [#47: List option should respect output format.](https://github.com/haensl/js-performance/issues/47)
* Add option to specify unit.
* Add documentation.

### 1.10.0
* [#11: Add profiles to compare Map vs. object literal.](https://github.com/haensl/js-performance/issues/11)

### 1.9.0
* [#32: Add JSON reporter.](https://github.com/haensl/js-performance/issues/32)
* Major refactoring of project file structure

### 1.8.0
* [#10: Add profile for guards.](https://github.com/haensl/js-performance/issues/10)
* Add command line option to specify result precision.
* Add test data type `object`.

### 1.7.0
* [#25: Add command line option to list available profiles.](https://github.com/haensl/js-performance/issues/25)

### 1.6.2
* [#23: DRY: add global `requireModule` function.](https://github.com/haensl/js-performance/issues/23)

### 1.6.1
* [#8: Rename profiles/all to profiles/index.js.](https://github.com/haensl/js-performance/issues/8)

### 1.6.0
* [#1: Update readme.](https://github.com/haensl/js-performance/issues/1)
* Adjust verbosity settings.
* Update project description.

### 1.5.0
* [#13: Separation of concerns.](https://github.com/haensl/js-performance/issues/13)

### 1.4.0
* [#22: Add TravisCI integration.](https://github.com/haensl/js-performance/issues/22)

### 1.3.1
* [#21: Cleanup eslint.](https://github.com/haensl/js-performance/issues/21)

### 1.3.0
* [#6: Add command line option to specify magnitude of test data.](https://github.com/haensl/js-performance/issues/6)

### 1.2.0
* [#4: Add recursion profile.](https://github.com/haensl/js-performance/issues/4)

### 1.1.0
* [#2: Add unit tests.](https://github.com/haensl/js-performance/issues/2)
* Refactor profiles.
* Refactor `performance` module into `clock` and `testdata`.

### 1.0.0
* Initial project setup.


js-performance.js
#!/usr/bin/env node --expose-gc

const glob = require('glob');
const join = require('path').join;
const GetOpt = require('node-getopt');
const chalk = require('chalk');
const DEFAULTS = require('./lib/support/defaults');
const VERBOSITY = require('./lib/support/verbosity');
const UNITS = require('./lib/support/units');
const jsperformance = require('./lib');

const opts = new GetOpt([
    ['h', 'help', 'Display this helptext.'],
    ['i', 'iterations=', `Specify the number of iterations per profiled function. Default: ${DEFAULTS.iterations}.`],
    ['j', 'json', `Output results in JSON format.`],
    ['l', 'list', 'List available profiles.'],
    ['m', 'magnitude=', `Specify the magnitude of testdata. Default: ${DEFAULTS.magnitude}.`],
    ['', 'memory', 'If present, memory consumption is measured.'],
    ['p', 'precision=', `Specify the precision in terms of decimal places of results. Default: ${DEFAULTS.precision.time},${DEFAULTS.precision.memory}. Separate time and memory precision by comma.`],
    ['q', 'quiet', 'Print results only.'],
    ['u', 'unit=', `Specify the unit for time and memory output. Default: ${DEFAULTS.units.time},${DEFAULTS.units.memory}. Possible values: auto (automatically convert between milli- and microseconds), ms (milliseconds), µs (microseconds), B (Bytes), KB (kilobytes), MB (megabytes). Separate time and memory unit by comma.`],
    ['v', 'verbose', 'Print verbose information.']
  ]).setHelp(
    'Usage: js-performance [OPTIONS] [profile1 profile2 ...]\n\n' +
    'Parameters:\n' +
    '  profile1 profile2 ...\n' +
    '  Optionally specify the profiles you want to run separated by spaces.\n\n' +
    'Options:\n' +
    '[[OPTIONS]]\n'
  ).bindHelp().parseSystem();

const options = Object.assign({}, DEFAULTS, { console: true });

if ('quiet' in opts.options) {
  options.verbosity = VERBOSITY.QUIET;
}

if ('verbose' in opts.options) {
  options.verbosity = VERBOSITY.VERBOSE;
}

if ('json' in opts.options) {
  options.json = true;
  options.console = false;
}

if ('list' in opts.options) {
  const profileList = jsperformance.list();
  if (options.json) {
    console.info(JSON.stringify(profileList.map((p) => {
      switch(options.verbosity) {
        case VERBOSITY.VERBOSE:
          return Object.assign({}, p, { description: p.description.long });
        case VERBOSITY.NORMAL:
          return {
            name: p.name,
            description: p.description.short
          };
      }
    }).filter((p) => typeof p === 'object'), null, 2));
  } else {
    console.info(profileList.map((profile) => {
      switch (options.verbosity) {
        case VERBOSITY.VERBOSE:
          return `${chalk.bold.underline(profile.name)}\n` +
            `${profile.description.long}\n` +
            `Profiled functions:\n` +
            `${profile.functions.map((f) => `  ${f}`)
              .join('\n')}`
        case VERBOSITY.NORMAL:
          return `${chalk.bold.underline(profile.name)}: ${profile.description.short}`;
      }
    }).filter((p) => typeof p === 'string').join('\n\n'));
  }

  process.exit(0);
}

if ('iterations' in opts.options) {
  if (!isNaN(parseInt(opts.options.iterations, 10))) {
    options.iterations = parseInt(opts.options.iterations, 10);
  } else {
    console.warn(chalk.yellow(`WARNING: "${opts.options.iterations}" is not a valid iterations number. Defaulting to ${options.iterations}.`));
  }
}

if ('magnitude' in opts.options) {
  if (!isNaN(parseInt(opts.options.magnitude, 10))) {
    options.magnitude = parseInt(opts.options.magnitude, 10);
  } else {
    console.warn(chalk.yellow(`WARNING: "${opts.options.magnitude}" is not a valid magnitude number. Defaulting to ${options.magnitude}.`));
  }
}

if ('precision' in opts.options) {
  const precision = opts.options.precision.split(',');
  if (!isNaN(parseInt(precision[0], 10))) {
    options.precision.time = precision[0];
  } else {
    console.warn(chalk.yellow(`WARNING: "${precision[0]}" is not a valid time precision. Defaulting to ${options.precision.time}.`));
  }

  if (precision.length > 1) {
    if (!isNaN(parseInt(precision[1], 10))) {
      options.precision.memory = precision[1];
    } else {
      console.warn(chalk.yellow(`WARNING: "${precision[1]}" is not a valid memory precision. Defaulting to ${options.precision.memory}.`));
    }
  }
}

if ('unit' in opts.options) {
  const units = opts.options.unit.split(',');
  if (UNITS.isValidTimeUnit(units[0])) {
    options.units.time = units[0];
  } else {
    console.warn(chalk.yellow(`WARNING: "${units[0]}" is not a valid time unit. Defaulting to ${options.units.time}.`));
  }
  
  if (units.length > 1) {
    if (UNITS.isValidMemoryUnit(units[1])) {
      options.units.memory = units[1];
    } else {
      console.warn(chalk.yellow(`WARNING: "${units[1]}" is not a valid memory unit. Defaulting to ${options.units.memory}.`));
    }
  }
}

if ('memory' in opts.options) {
  options.memory = true;
}

if (opts.argv.length > 0) {
  options.profiles = opts.argv;
}

jsperformance.run(options)
  .catch((err) => {
    console.error(err.stack);
    process.exit(1);
  });
